File: SooBin/231225_week/Baekjoon_1916_최소비용 구하기.cpp
#include<iostream>
#include<algorithm>
#include<vector>
#include<queue>
#include<climits>

using namespace std;

vector<pair<int, int>> graph[1001];
long long d[1001];

void dijkstra(int start) {
    priority_queue<pair<int, int>> pq;
    pq.push({ 0,start });   
    d[start] = 0;
    while (!pq.empty()) {
        long long dist = -pq.top().first;
        int node = pq.top().second;
        pq.pop();

        // !
        if (d[node] < dist) continue;

        for (unsigned int i = 0; i < graph[node].size(); i++) {
            long long cost = dist + graph[node][i].second;

            // !
            if (cost < d[graph[node][i].first]) {
                d[graph[node][i].first] = cost;
                pq.push({ -cost,graph[node][i].first });
            }
        }

    }
}

int main() {
    
    int N, M;
    cin >> N >> M;

    int a, b, c;
    for (int i = 0; i < M; i++) {
        cin >> a >> b >> c;
        graph[a].push_back({ b,c });
    }

    int A, B;
    cin >> A >> B;

 
    fill(d, d + 1001, LLONG_MAX);
    dijkstra(A);

    cout << d[B];  return 0;
}

  

File: SooYeon/240212_week/Baekjoon_1916_최소비용 구하기.js
let fs = require('fs');
let input = fs.readFileSync('/dev/stdin').toString().split('\n');

function PriorityQueue(comparator) {
  this._comparator = comparator || PriorityQueue.DEFAULT_COMPARATOR;
  this._elements = [];
}

PriorityQueue.DEFAULT_COMPARATOR = function(a, b) {
  if (typeof a === 'number' && typeof b === 'number') {
    return a - b;
  } else {
    a = a.toString();
    b = b.toString();

    if (a == b) return 0;

    return (a > b) ? 1 : -1;
  }
};

PriorityQueue.prototype.isEmpty = function() {
  return this.size() === 0;
};

PriorityQueue.prototype.peek = function() {
  if (this.isEmpty()) throw new Error('PriorityQueue is empty');

  return this._elements[0];
};

PriorityQueue.prototype.deq = function() {
  var first = this.peek();
  var last = this._elements.pop();
  var size = this.size();

  if (size === 0) return first;

  this._elements[0] = last;
  var current = 0;

  while (current < size) {
    var largest = current;
    var left = (2 * current) + 1;
    var right = (2 * current) + 2;

    if (left < size && this._compare(left, largest) >= 0) {
      largest = left;
    }

    if (right < size && this._compare(right, largest) >= 0) {
      largest = right;
    }

    if (largest === current) break;

    this._swap(largest, current);
    current = largest;
  }

  return first;
};

PriorityQueue.prototype.enq = function(element) {
  var size = this._elements.push(element);
  var current = size - 1;

  while (current > 0) {
    var parent = Math.floor((current - 1) / 2);

    if (this._compare(current, parent) <= 0) break;

    this._swap(parent, current);
    current = parent;
  }

  return size;
};

PriorityQueue.prototype.size = function() {
  return this._elements.length;
};

PriorityQueue.prototype.forEach = function(fn) {
  return this._elements.forEach(fn);
};

PriorityQueue.prototype._compare = function(a, b) {
  return this._comparator(this._elements[a], this._elements[b]);
};

PriorityQueue.prototype._swap = function(a, b) {
  var aux = this._elements[a];
  this._elements[a] = this._elements[b];
  this._elements[b] = aux;
};

// let input = `5
// 8
// 1 2 2
// 1 3 3
// 1 4 1
// 1 5 10
// 2 4 2
// 3 4 1
// 3 5 1
// 4 5 3
// 1 5`.split('\n');

let INF = 1e9;
let n = Number(input[0]);
let m = Number(input[1]);
let graph = Array.from({length: n + 1}, () => []);
let distance = new Array(n + 1).fill(INF);

for(let i = 2; i < m + 2; i++) {
  let [start, end, weight] = input[i].split(' ').map(Number);
  graph[start].push([end, weight]);
}

let [start, end] = input[m + 2].split(' ').map(Number);

const dijkstra = (start) => {
  const pq = new PriorityQueue((a, b) => b[0] - a[0]);

  pq.enq([0, start]);
  distance[start] = 0;

  while(pq.size()) {
    const [dist, now] = pq.deq();

    if(distance[now] < dist) continue;

    for(let i of graph[now]) {
      let cost = dist + i[1];

      if(cost < distance[i[0]]) {
        distance[i[0]] = cost;
        pq.enq([cost, i[0]]);
      
      }
    }
  }
}

dijkstra(start);
console.log(distance[end]);